docs/address_lookup/the_address_lookup_process.md

```diff
@@ -0,0 +1,95 @@
+
+| OS Places API Structure |
+|-------------------------|
+| UPRN                    |
+| SAO_START_NUMBER        |
+| SAO_START_SUFFIX        |
+| SAO_END_NUMBER          |
+| SAO_END_SUFFIX          |
+| SAO_TEXT                |
+| PAO_START_NUMBER        |
+| PAO_START_SUFFIX        |
+| PAO_END_NUMBER          |
+| PAO_END_SUFFIX          |
+| PAO_TEXT                |
+| ORGANISATION            |
+| STREET_DESCRIPTION      |
+| LOCALITY_NAME           |
+| TOWN_NAME               |
+| ADMINISTRATIVE_AREA     |
+| AREA_NAME               |
+| POSTCODE_LOCATOR        |
+
+Stored addresses are required to be in the this structure:
+
+| Required Structure         |
+|----------------------------|
+| uprn                       |
+| building_and_street_line_1 |
+| building_and_street_line_two |
+| town_city                  |
+| county                     |
+| postcode                   |
+
+As the two structures differ significantly, conversion is required between the two.  The following diagram describes the conversion logic.
+
+![Address conversion](images/address_conversion.png)
+
+<p><center><strong>Address conversion</strong></center></p>
+
+## Primary and Secondary address objects
+
+The OS Places API contains two blocks of sub-address elements: one set beginning `PAO_` and the other set beginning `SAO_`.  These identify the Primary Address Object (PAO) and the Secondary Address Object (SAO). All of these fields are optional for an address.
+
+The easiest way to understand how these data objects function is to briefly describe the delivery process:
+
+  * The first part of the postcode - the Outward code (e.g. __E1__) is used to route the mail to the correct postal delivery office.
+
+  * The second part of the postcode - the Inward code (e.g. __8QS__) is used tell the postal delivery office which delivery round the address is on (e.g. __Whitechapel High Street__).
+
+  * There can be many buildings within a single delivery round, so the PAO is used to identify a specific building - the primary address (e.g. __10 Whitechapel High Street__).
+
+  * Finally, the SAO is used to identify the area within the primary address (e.g. __Office Suite 5__).
+
+Internally, we concatenate the individual elements of the PAO into a single field (called `PAO`). The same is done with the SAO - with the addition of the `ORGANISATION` field.
+
+## Comparison
+
+The user is allowed to edit the converted address. We need to retain the UPRN against any changes the user might make. In order to do so, we need to compare the known - selected - address with the results of the users edit.
+
+The logic for doing this comparison can be boiled down into the following...
+
+```ruby
+if SAO.present?
+  known_address.lines_one_two_and_postcode == edited_address.lines_one_two_and_postcode
+else
+  known_address.lines_one_two_and_postcode == edited_address.lines_one_two_and_postcode ||
+  known_address.line_one_and_postcode == edited_address.line_one_and_postcode
```

Comment:
These need editing to remove the "and" and will need changing when we start to look at the bug

Reply:
Thanks Jon.  So, this is more pseudo-code than actual code - I totally get it that my intent is not clear.  Sorry.  I'll see if I can make my intent clearer.

Reply:
Actually I think I just misread something, this works for how it is at the moment, I wouldn't worry about changes now. I do think that it'll need to change once we have a look at the bug but that shouldn't be an issue given that one of us will probably be working on that.